### ISSUE TITLE
Trio support

### ISSUE THREAD

**Issue by jordaneremieff** [jordaneremieff/mangum#24]
So I admittedly don't know how much may be involved here, but my initial thoughts are that including support for ASGI applications using Trio wouldn't require too much to implement as an adapter here.

**Comment by tomchristie** [comment]
So... The only place you're using `asyncio` is for a single Queue instance. For trio, you'd need to swap that out with a trio Queue implementation instead.

The ASGI interface itself is agnostic - it's an `async`/`await` interface, rather than specifically being an `asyncio` interface.

(Broader context, Trio is obviously a *huge* step forwards, but I'm *personally* lukewarm on the ecosystem split it requires vs. bringing the learnings it introduces into asyncio.)

**Comment by erm** [comment]
@tomchristie Hm, interesting point about the broader context. I haven't really delved into Trio too deeply, but I've only heard good things about it and have seen various efforts to use it with ASGI.

That said, I don't have any particular interest in using it myself. I'm going to leave this as a "maybe" and won't likely do anything further, but if there are enough people that would find an ASGI Trio adapter useful I would be open to a PR.

**Comment by tomchristie** [comment]
Codewise I think it just needs a trivial change to use `trio.Event()` in `mangum/asgi/protocol.py` (Tho I'm not sure how best you'd expose that as a switch.) Anyways, leaving this here, for someone else to make progress on if it ends up being a requirement for someone, somewhere

**Comment by erm** [comment]
Starlette is considering support for Trio [here](https://github.com/encode/starlette/issues/811) and that would make at least two frameworks that support Trio (Quart being the other) - might be worth including here.

I'll reopen this if Trio support lands in Starlette.

**Comment by jmehnle** [comment]
Starlette ended up implementing Trio support by migrating onto AnyIO in September 2022. Any thoughts on the way forward for Mangum?

**Comment by jmehnle** [comment]
@tomchristie, I'm not familiar with the Mangum implementation, why did you suggest to use `trio.Event` in `mangum/asgi/protocol.py` above? Was that meant as a substitute for the use of `asyncio.Queue` (now in `mangum/protocols/http.py`), or is there an _additional_ synchronization problem?

**Comment by tomchristie** [comment]
> IMO the ideal solution for Trio support would be to make Mangum use AnyIO

Yep. Tho... I don't really understand enough if there's a fundamental incompatibility between structured concurrency and the ASGI protocol. It's feasible that you might start work on this and then run into issues when translating how the queues used to work onto memory channels??? (Or it might be dead easy, and just needs updating to the more recent primitives)

> why did you suggest to use `trio.Event` in `mangum/asgi/protocol.py` above?

I don't recall exactly and timed out while digging through the history just now, trying to reference it.
